### DISCUSSION TITLE
Support for C++20

### DISCUSSION THREAD

**Discussion by TorstenRobitzki** [zephyrproject-rtos/zephyr#68130] [Q&A]
Hi,
I would love to use the C++20 features of the compiler I'm using (gcc 12.2.1). The relevant parts of my `prj.conf` looks like this:

    # Enable support for C++
    CONFIG_CPP=y
    CONFIG_REQUIRES_FULL_LIBCPP=y
    CONFIG_NEWLIB_LIBC=y
    CONFIG_STD_CPP20=y

However, when I compile C++ code, there is no switch passed to the compiler, selecting `C++20`. When I add: 

    set(CMAKE_CXX_STANDARD 20)
    set(CMAKE_CXX_STANDARD_REQUIRED ON)

to the CMakeLists.txt file, I get the required compiler switch (-std=gnu++20). I wonder, if this is really necessary.

The second thing is, that when I load of some sources into a static library target and add them as `target_link_library` to the `app` target, I get error when linking:

    [ 97%] Linking CXX executable zephyr_pre0.elf
    /usr/local/arm-gnu-toolchain-12.2.rel1/bin/../lib/gcc/arm-none-eabi/12.2.1/../../../../arm-none-eabi/bin/ld.bfd: error: ../source/libcobs.a(cobs.cpp.obj): conflicting CPU architectures 17/2
    /usr/local/arm-gnu-toolchain-12.2.rel1/bin/../lib/gcc/arm-none-eabi/12.2.1/../../../../arm-none-eabi/bin/ld.bfd: failed to merge target specific data of file ../source/libcobs.a(cobs.cpp.obj)
    /usr/local/arm-gnu-toolchain-12.2.rel1/bin/../lib/gcc/arm-none-eabi/12.2.1/../../../../arm-none-eabi/bin/ld.bfd: warning: zephyr_pre0.elf has a LOAD segment with RWX permissions
    collect2: error: ld returned 1 exit status

When I'm adding the sources to the `app` target, that I do not see this linker errors.

What am I'm doing wrong here?

Thanks in advance,

Torsten

**Comment by KammutierSpule** [comment]
Any update on this?
I also get this error: conflicting CPU architectures 17/2

**Comment by TorstenRobitzki** [comment]
I gave up on this as the purpose of having libraries is usually to link them to different executables. As Zephyr seems to not support multiple executables, I came up with this workaround: https://github.com/zephyrproject-rtos/zephyr/discussions/69029

Which feels totally wrong, as CMake has all the required features and then, you throw away everything and build some crapy replacement on top of it.

OBJECT libraries could work, as they are not really libraries, but lists of object files that are handled by CMake.
